perses/annihilation/relative.py

```diff
@@ -2692,39 +2654,197 @@ def __init__(self,
         # Generate the topology representation
         self._hybrid_topology = self._create_topology()
 
-        # define the scale and nonscale region
-        self._make_scale_regions(scale_region)
+    def _handle_scale_regions(self, scale_regions):
+        num_scale_regions = None
+        if scale_regions is None:
+            self._num_scale_regions = 0
+            self._scale_regions = None
+        else:
+            assert type(scale_regions) == list
+            num_scale_regions = len(scale_regions)
+            for entry in scale_regions:
+                assert type(entry) == list
+
+            #the last bit is to ensure that the scale regions are disjoint
+            num_scaled_particles = sum([len(i) for i in scale_regions])
+            import itertools
+            num_combined_scaled_particles = len(list(itertools.chain.from_iterable(scale_regions)))
+            assert num_scaled_particles == num_combined_scaled_particles, f"each scale region must be disjoint"
+            self._scale_regions = [set(i) for i in scale_regions]
+
+        #get the scale template
+        self._scale_templates = [
+                                    list(itertools.chain.from_iterable(['1'] + [[f"scale_lambda_{i}", f"interscale_lambda_{i}"] for i in range(self._num_scale_regions)])),
+                                    list(itertools.chain.from_iterable(['nonscale_region'] + [[f"scale_region_{i}", f"interscale_region_{i}"] for i in range(self._num_scale_regions)]))
+        ]
+
+
+    def _determine_atom_classes(self):
+        """
+        This method determines whether each atom belongs to unique old, unique new, core, or environment, as defined above.
+        All the information required is contained in the TopologyProposal passed to the constructor. All indices are
+        indices in the hybrid system.
+
+        Returns
+        -------
+        atom_classes : dict of list
+            A dictionary of the form {'core' :core_list} etc.
+        """
+        from itertools import chain
+
+        is_single_alch_region = True if type(self._topology_proposal._core_new_to_old_atom_map) == dict else False
+        if is_single_alch_region:
+            assert type(self._topology_proposal._unique_old_atoms) == list
+            assert type(self._topology_proposal._unique_new_atoms) == list
+            self._num_alchemical_regions = 1
+        else:
+            len_list =len(self._topology_proposal._core_new_to_old_atom_map)
+            for entry in [self._topology_proposal._core_new_to_old_atom_map, self._topology_proposal._unique_old_atoms, self._topology_proposal._unique_new_atoms]:
+                assert entry == list
+                assert len(entry) == len_list
```

Comment:
are you sure that you want every element in the loop to have the same length as the `core_new_to_old_atom_map`?